state/upgrades.go

```
@@ -65,185 +41,153 @@ func runForAllEnvStates(st *State, runner func(st *State) error) error {
+func stripLocalFromFields(st *State, collName string, fields ...string) ([]txn.Op, error) {
+	coll, closer := st.getRawCollection(collName)
 	defer closer()
+	upgradesLogger.Infof("migrating document fields of the %s collection", collName)
 
-	upgradesLogger.Debugf("migrating schema of the %s collection", settingsC)
 	iter := coll.Find(nil).Iter()
 	defer iter.Close()
-
 	var ops []txn.Op
-	var doc bson.M
+	var doc bson.D
 	for iter.Next(&doc) {
-		if !settingsDocNeedsMigration(doc) {
-			continue
+		// Get a copy of the current doc id so we can see if it has changed.
+		id, err := readBsonDField(doc, "_id")
+		if err != nil {
+			return nil, err
 		}
+		newId := id
 
-		id := doc["_id"]
-		txnRevno := doc["txn-revno"].(int64)
-
-		// Remove reserved attributes; we'll move the remaining
-		// ones to the "settings" subdoc.
-		delete(doc, "model-uuid")
-		delete(doc, "_id")
-		delete(doc, "txn-revno")
-		delete(doc, "txn-queue")
-
-		// If there exists a setting by the name "settings",
-		// we must remove it first, or it will collide with
-		// the dotted-notation $sets.
-		if _, ok := doc["settings"]; ok {
-			ops = append(ops, txn.Op{
-				C:      settingsC,
-				Id:     id,
-				Assert: txn.DocExists,
-				Update: bson.D{{"$unset", bson.D{{"settings", 1}}}},
-			})
+		// Take a copy of the current doc fields.
+		newDoc := make(bson.D, len(doc))
+		for i, f := range doc {
+			newDoc[i] = f
 		}
 
+		// Iterate over the fields that need to be updated and
+		// record any updates to be made.
 		var update bson.D
-		for key, value := range doc {
-			if key != "settings" && key != "version" {
-				// Don't try to unset these fields,
-				// as we've unset "settings" above
-				// already, and we'll overwrite
-				// "version" below.
-				update = append(update, bson.DocElem{
-					"$unset", bson.D{{key, 1}},
-				})
+		for _, field := range fields {
+			isId := field == "_id"
+			fieldVal, err := readBsonDField(doc, field)
+			if err != nil {
+				continue
```

it's a bit alarming to ignore the error. can we either check for NotFound specifically, or otherwise change readBsonField to return a bool instead of an error?